File: mobile/lib/features/menu/domain/current_week.dart
/// UTC midnight of [date]'s own calendar-date components (year/month/day),
/// discarding [date]'s own time-of-day and timezone. The one canonical
/// normalizer for "a calendar date, represented as the UTC-midnight
/// `DateTime` this codebase's `Menu`/`MenuKey` machinery expects" — reused by
/// [currentWeekStartDate] below and by `state/current_menu_controller.dart`'s
/// `menuKeyFor` (see that function's own doc for why this normalization
/// matters: `DateTime.==` considers `isUtc` and microseconds, not just the
/// calendar date, so two un-normalized `DateTime`s meaning the same week
/// would otherwise compare unequal).
DateTime utcMidnight(DateTime date) =>
    DateTime.utc(date.year, date.month, date.day);

/// The Monday (`Menu.weekStartDate`, `dayOfWeek: 0`) of the week containing
/// [today] — computed client-side, from the DEVICE's own local calendar
/// date, never server-computed (E2E_MVP_PLAN.md §15.2.4's locked decision:
/// "today"/week boundaries are calendar dates, deliberately not
/// timezone-attached, and there is no household-level timezone column to
/// compute them against server-side even if this codebase wanted to).
///
/// [today] defaults to [DateTime.now()] — a parameter (not a hardcoded call
/// inside the body) purely so a test can pass a fixed date instead of racing
/// the real clock.
DateTime currentWeekStartDate({DateTime? today}) {
  final DateTime now = today ?? DateTime.now();
  // `Menu.dayOfWeek: 0` is the week's first day (see `Menu.itemsForDay`'s own
  // doc) — this codebase's own migration/test precedent treats that as
  // Monday (E2E_MVP_PLAN.md's own worked S1/S2 examples all place a Monday
  // at day 0). `DateTime.weekday` is already 1 (Monday) through 7 (Sunday),
  // so subtracting `weekday - 1` days lands on that same Monday.
  final DateTime monday = now.subtract(Duration(days: now.weekday - 1));
  return utcMidnight(monday);
}

File: web/src/domain/currentWeek.ts
/**
 * W18 S4's TypeScript mirror of `mobile/lib/features/menu/domain/
 * current_week.dart`'s `currentWeekStartDate`/`utcMidnight` — the dashboard
 * must compute "this week" via the identical rule mobile already uses
 * (E2E_MVP_PLAN.md §24's own "mirror the existing convention, don't invent a
 * second date-math rule"), not a new interpretation. `Menu.dayOfWeek: 0` is
 * Monday (confirmed against `weekdayNames`/`_DaySection` in
 * `weekly_plan_screen.dart`), and "today" is computed client-side from the
 * caller's own local calendar date, never server-computed
 * (E2E_MVP_PLAN.md §15.2.4's locked decision) — on the web, "client-side"
 * means the Node server process rendering the page, which is this function's
 * own `today` parameter, defaulting to `new Date()`.
 */

/** UTC midnight of `date`'s own calendar-date components, discarding its time-of-day and timezone. */
export const utcMidnight = (date: Date): Date =>
  new Date(Date.UTC(date.getFullYear(), date.getMonth(), date.getDate()));

/**
 * The Monday (`Menu.weekStartDate`, `dayOfWeek: 0`) of the week containing
 * `today`. `Date.getDay()` is 0 (Sunday) through 6 (Saturday) — converted to
 * the 1 (Monday) through 7 (Sunday) convention `current_week.dart`'s own
 * `DateTime.weekday` uses before applying the identical `weekday - 1` days
 * subtraction.
 */
export const currentWeekStartDate = (today: Date = new Date()): Date => {
  const isoWeekday = today.getDay() === 0 ? 7 : today.getDay();
  const monday = new Date(today);
  monday.setDate(today.getDate() - (isoWeekday - 1));
  return utcMidnight(monday);
};

/**
 * `currentWeekStartDate` as the `AWSDateTime` string `Query.menu`'s
 * `weekStartDate` argument expects (UTC midnight, ISO-8601).
 */
export const currentWeekStartDateIso = (today: Date = new Date()): string =>
  currentWeekStartDate(today).toISOString();
